README.md
# aclients
基于sanic扩展，各种数据库异步crud操作，各种请求异步crud操作,本repo是个基础封装库，直接用于各个业务系统中

# Installing aelog
- ```pip install aclients```

# Usage
### 后续添加，现在没时间.


CHANGES.md
## aclients Changelog

###[1.0.3] - 2020-10-13

#### Changed 
- 暂时去掉一个账户只能一处登录的限制
- session的有效期由30分钟延长到1个小时
- 上下文管理器改为异步上下文管理器

###[1.0.1] - 2020-6-14

#### Added 
- 增加生成分表model功能，使得分表的使用简单高效
- 配置增加pool_recycle回旋关闭连接功能

#### Changed 
- 优化所有代码中没有类型标注的地方,都改为typing中的类型标注
- 经过项目测试功能稳定,发布正式版本.
- 删除未开发完成的功能.
- 此扩展原则上将做为最后一个未拆分扩展版本,后续将分开维护.


###[1.0.0b45] - 2019-12-30

#### Added 
- 新增如果查询(分页查询)的时候没有进行排序，默认按照id关键字升序排序的功能,防止出现混乱数据

###[1.0.0b44] - 2019-9-29

#### Added 
- 增加incrbynumber方法，可以增加整数或者浮点数
- 添加四个扩展到app的反向引用
- 增加add_task异步任务执行

#### Changed 
- 优化redis client中所有的设置过期时间的实现方式
- 优化获取数据的方式，去掉不适宜的异常
- 更改session可能存在的隐藏问题

###[1.0.0b40] - 2019-9-11

#### Changed 
- 更改由于session过期后再次登录删除老的session报错的问题

###[1.0.0b39] - 2019-9-10

#### Added 
- 增加新的账号登录生成session后,清除老的session的功能,保证一个账号只能一个终端登录

#### Changed 
- 更改session的过期时间为30分钟
- 更改通用数据的缓存时间为12小时
- 删除session时删除和session所有相关的缓存key

###[1.0.0b37] - 2019-7-12

#### Changed 
- 更改schema valication的中文提示信息
- 更改http,mongo,mysql,redis中停止服务时会出现关闭pool报错的情况

###[1.0.0b36] - 2019-5-23

#### Added 
- 工具类中增加生成随机长度以字母开头的字母和数字的字符串标识

###[1.0.0b35] - 2019-4-29

#### Changed 
- 修改三方库的pymongo>=3.8.0,以上的版本实现了ObjectID 0.2版本规范,生成的ObjectID发生碰撞的可能行更小.

###[1.0.0b34] - 2019-4-24

#### Added 
- 修改aiohttp client增加针对ip地址的URL中session接受cookie的功能开关

###[1.0.0b33] - 2019-4-23

#### Added 
- 工具类utils中增加用于枚举实例的元类
- tinylibs包中增加简单的异步信号实现blinker模块

#### Changed 
- 移动单例类Singleton从decorators到utils工具类

###[1.0.0b32] - 2019-4-22

#### Changed 
- 修改schema message装饰器判断错误的情况


###[1.0.0b31] - 2019-4-21

#### Changed 
- 修改sanic版本为长期支持版本18.12LTS

###[1.0.0b30] - 2019-4-20

#### Changed 
- 优化redis客户端，修改出现的错误

###[1.0.0b29] - 2019-4-19

#### Changed 
- 修改redis中获取session返回session对象出错的问题

###[1.0.0b28] - 2019-4-19

#### Added 
- 增加保存和更新hash数据时对单个键值进行保存和更新的功能

###[1.0.0b27] - 2019-4-19

#### Changed 
- 修改获取redis数据时可能出现的没有把字符串转换为对象的情况
- 修改保存redis数据时指定是否进行dump以便进行性能的提高
- 修改获取redis数据时指定是否进行load以便进行性能的提高


###[1.0.0b26] - 2019-4-18

#### Changed 
- 修改Session中增加page_id和page_menu_id两个用于账户的页面权限管理

###[1.0.0b25] - 2019-4-16

#### Changed 
- 修改TinyMySQL中execute的参数，修改find中的args参数名称

###[1.0.0b24] - 2019-4-2

#### Added 
- 工具类中增加由对象名生成类名的功能
- 工具类中增加解析yaml文件的功能
- 工具类中增加返回objectid的功能
- 增加基于pymysql的简单TinyMySQL功能，用于简单操作MySQL的时候使用 

###[1.0.0b22] - 2019-3-25

#### Changed 
- 修改update data中设置前缀的错误

###[1.0.0b20] - 2019-3-22

#### Changed 
- 修改mongo插入中的错误，修改id时的错误

###[1.0.0b19] - 2019-3-22

#### Changed 
- 修改mongo查询中的错误

###[1.0.0b18] - 2019-3-21

#### Added 
- redis的session中增加角色ID
- redis的session中增加静态权限ID
- redis的session中增加动态权限ID

#### Changed 
- redis的session中user_id更改为account_id

###[1.0.0b17] - 2019-3-21

#### Added 
- 增加同步方法包装为异步方法的功能 

###[1.0.0b16] - 2019-3-16

#### Changed 
- 优化insert document中的ID处理逻辑 
- 优化update data处理逻辑 
- 优化query key处理逻辑，可以直接使用in、like等查询

###[1.0.0b15] - 2019-1-31

#### Changed 
- 修改schema装饰器实现

###[1.0.0b14] - 2019-1-31

#### Added 
- 增加元类单例实现

#### Changed 
- 修改httpclient为元类单例的子类

###[1.0.0b13] - 2019-1-28

#### Changed 
- 优化MySQL或查询支持列表

###[1.0.0b12] - 2019-1-28

#### Changed 
- 删除http message 中不用的消息
- 优化exceptions的实现方式

###[1.0.0b11] - 2019-1-25

#### Changed 
- 修改schema_validate装饰器能够修改提示消息的功能,如果多个地方用到此装饰器，在其中一处修改即可

###[1.0.0b10] - 2019-1-25

#### Added 
- 添加schema_validate装饰器用于校验schema
#### Changed 
- 修改http message默认值

###[1.0.0b9] - 2019-1-21

#### Added 
- 增加多数据库、多实例应用方式
- 增加http client的测试
- 增加在没有app时脚本中使用时的初始化功能,这样便于通用性
- 增加错误类型，能够对错误进行定制
- 增加单例的装饰器，修改httpclient为单例
#### Changed 
- 修改一处可能引起错误的地方

###[1.0.0b8] - 2019-1-18

#### Changed 
- 修改初始化方式，更改为sanic扩展的初始化方式，即init_app
- 修改初始化时配置的加载顺序，默认先加载

###[1.0.0b7] - 2019-1-18

#### Changed 
- 从b2到b7版本的修改记录忘记了，这里先不记录了

###[1.0.0b1] - 2018-12-26

#### Added 

- MySQL基于aiomysql和sqlalchemy的CRUD封装
- http基于aiohttp的CRUD封装
- session基于aredis的CRUD封装
- redis基于aredis的常用封装
- mongo基于motor的CRUD封装
- 所有消息可自定义配置,否则为默认配置


aclients/aio_redis_client.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-12-25 下午5:15
"""
import atexit
import secrets
import uuid
from collections import MutableMapping
from typing import Any, Dict, List, NoReturn, Union

import aelog
import aredis
import ujson
from aredis import RedisError

from .exceptions import RedisClientError
from .utils import async_ignore_error

__all__ = ("Session", "AIORedisClient")

LONG_EXPIRED: int = 24 * 60 * 60  # 最长过期时间
EXPIRED: int = 12 * 60 * 60  # 通用过期时间
SESSION_EXPIRED: int = 60 * 60  # session过期时间


class Session(object):
    """
    保存实际看结果的session实例
    Args:

    """

    def __init__(self, account_id: str, *, session_id: str = None, org_id: str = None, role_id: str = None,
                 permission_id: str = None, **kwargs):
        self.account_id = account_id  # 账户ID
        self.session_id = secrets.token_urlsafe() if not session_id else session_id  # session ID
        self.org_id = org_id or uuid.uuid4().hex  # 账户的组织结构在redis中的ID
        self.role_id = role_id or uuid.uuid4().hex  # 账户的角色在redis中的ID
        self.permission_id = permission_id or uuid.uuid4().hex  # 账户的权限在redis中的ID
        self.static_permission_id = uuid.uuid4().hex  # 账户的静态权限在redis中的ID
        self.dynamic_permission_id = uuid.uuid4().hex  # 账户的动态权限在redis中的ID
        self.page_id = uuid.uuid4().hex  # 账户的页面权限在redis中的ID
        self.page_menu_id = uuid.uuid4().hex  # 账户的页面菜单权限在redis中的ID
        for k, v in kwargs.items():
            setattr(self, k, v)


class AIORedisClient(object):
    """
    redis 非阻塞工具类
    """

    def __init__(self, app=None, *, host: str = "127.0.0.1", port: int = 6379, dbname: int = 0, passwd: str = "",
                 pool_size: int = 50):
        """
        redis 非阻塞工具类
        Args:
            app: app应用
            host:redis host
            port:redis port
            dbname: database name
            passwd: redis password
            pool_size: redis pool size
        """
        self.app = app
        self.pool = None
        self.redis_db: aredis.StrictRedis = None
        self.host = host
        self.port = port
        self.dbname = dbname
        self.passwd = passwd
        self.pool_size = pool_size
        self._account_key = "account_to_session"

        if app is not None:
            self.init_app(app, host=self.host, port=self.port, dbname=self.dbname, passwd=self.passwd,
                          pool_size=self.pool_size)

    def init_app(self, app, *, host: str = None, port: int = None, dbname: int = None, passwd: str = "",
                 pool_size: int = None):
        """
        redis 非阻塞工具类
        Args:
            app: app应用
            host:redis host
            port:redis port
            dbname: database name
            passwd: redis password
            pool_size: redis pool size
        Returns:

        """
        self.app = app
        host = host or app.config.get("ACLIENTS_REDIS_HOST", None) or self.host
        port = port or app.config.get("ACLIENTS_REDIS_PORT", None) or self.port
        dbname = dbname or app.config.get("ACLIENTS_REDIS_DBNAME", None) or self.dbname
        passwd = passwd or app.config.get("ACLIENTS_REDIS_PASSWD", None) or self.passwd
        pool_size = pool_size or app.config.get("ACLIENTS_REDIS_POOL_SIZE", None) or self.pool_size

        passwd = passwd if passwd is None else str(passwd)

        @app.listener('before_server_start')
        async def open_connection(app_, loop):
            """

            Args:

            Returns:

            """
            # 返回值都做了解码，应用层不需要再decode
            self.pool = aredis.ConnectionPool(host=host, port=port, db=dbname, password=passwd, decode_responses=True,
                                              max_connections=pool_size)
            self.redis_db = aredis.StrictRedis(connection_pool=self.pool, decode_responses=True)

        @app.listener('after_server_stop')
        async def close_connection(app_, loop):
            """
            释放redis连接池所有连接
            Args:

            Returns:

            """
            self.redis_db = None
            if self.pool:
                self.pool.disconnect()

    def init_engine(self, *, host: str = None, port: int = None, dbname: int = None, passwd: str = "",
                    pool_size: int = None):
        """
        redis 非阻塞工具类
        Args:
            host:redis host
            port:redis port
            dbname: database name
            passwd: redis password
            pool_size: redis pool size
        Returns:

        """
        host = host or self.host
        port = port or self.port
        dbname = dbname or self.dbname
        passwd = passwd or self.passwd
        pool_size = pool_size or self.pool_size

        passwd = passwd if passwd is None else str(passwd)
        # 返回值都做了解码，应用层不需要再decode
        self.pool = aredis.ConnectionPool(host=host, port=port, db=dbname, password=passwd, decode_responses=True,
                                          max_connections=pool_size)
        self.redis_db = aredis.StrictRedis(connection_pool=self.pool, decode_responses=True)

        @atexit.register
        def close_connection():
            """
            释放redis连接池所有连接
            Args:

            Returns:

            """
            self.redis_db = None
            if self.pool:
                self.pool.disconnect()

    async def save_session(self, session: Session, dump_responses: bool = False, ex: int = SESSION_EXPIRED) -> str:
        """
        利用hash map保存session
        Args:
            session: Session 实例
            dump_responses: 是否对每个键值进行dump
            ex: 过期时间，单位秒
        Returns:

        """
        session_data = await self.response_dumps(dump_responses, session)

        try:
            if not await self.redis_db.hmset(session_data["session_id"], session_data):
                raise RedisClientError("save session failed, session_id={}".format(session_data["session_id"]))
            if not await self.redis_db.expire(session_data["session_id"], ex):
                aelog.error("set session expire failed, session_id={}".format(session_data["session_id"]))
        except RedisError as e:
            aelog.exception("save session error: {}, {}".format(session.session_id, e))
            raise RedisClientError(str(e))
        else:
            # 清除老的令牌
            # try:
            #     old_session_id = await self.get_hash_data(self._account_key, field_name=session.account_id)
            # except RedisClientError as e:
            #     aelog.info(f"{session.account_id} no old token token, {str(e)}")
            # else:
            #     async with async_ignore_error():
            #         await self.delete_session(old_session_id, False)
            # # 更新新的令牌
            # await self.save_update_hash_data(self._account_key, field_name=session.account_id,
            #                                  hash_data=session.session_id, ex=LONG_EXPIRED)
            return session.session_id

    @staticmethod
    async def response_dumps(dump_responses: bool, session: Session) -> Dict:
        session_data = dict(vars(session))
        # 是否对每个键值进行dump
        if dump_responses:
            hash_data = {}
            for hash_key, hash_val in session_data.items():
                if not isinstance(hash_val, str):
                    async with async_ignore_error():
                        hash_val = ujson.dumps(hash_val)
                hash_data[hash_key] = hash_val
            session_data = hash_data
        return session_data

    async def delete_session(self, session_id: str, delete_key: bool = True) -> NoReturn:
        """
        利用hash map删除session
        Args:
            session_id: session id
            delete_key: 删除account到session的account key
        Returns:

        """

        try:
            session_id_ = await self.redis_db.hget(session_id, "session_id")
            if session_id_ != session_id:
                raise RedisClientError("invalid session_id, session_id={}".format(session_id))
            exist_keys = []
            session_data = await self.get_session(session_id, cls_flag=False)
            exist_keys.append(session_data["org_id"])
            exist_keys.append(session_data["role_id"])
            exist_keys.append(session_data["permission_id"])
            exist_keys.append(session_data["static_permission_id"])
            exist_keys.append(session_data["dynamic_permission_id"])
            exist_keys.append(session_data["page_id"])
            exist_keys.append(session_data["page_menu_id"])

            async with async_ignore_error():  # 删除已经存在的和账户相关的缓存key
                await self.delete_keys(exist_keys)
                # if delete_key is True:
                #     await self.redis_db.hdel(self._account_key, session_data["account_id"])

            if not await self.redis_db.delete(session_id):
                aelog.error("delete session failed, session_id={}".format(session_id))
        except RedisError as e:
            aelog.exception("delete session error: {}, {}".format(session_id, e))
            raise RedisClientError(str(e))

    async def update_session(self, session: Session, dump_responses: bool = False,
                             ex: int = SESSION_EXPIRED) -> NoReturn:
        """
        利用hash map更新session
        Args:
            session: Session实例
            ex: 过期时间，单位秒
            dump_responses: 是否对每个键值进行dump
        Returns:

        """
        session_data = await self.response_dumps(dump_responses, session)

        try:
            if not await self.redis_db.hmset(session_data["session_id"], session_data):
                raise RedisClientError("update session failed, session_id={}".format(session_data["session_id"]))
            if not await self.redis_db.expire(session_data["session_id"], ex):
                aelog.error("set session expire failed, session_id={}".format(session_data["session_id"]))
        except RedisError as e:
            aelog.exception("update session error: {}, {}".format(session_data["session_id"], e))
            raise RedisClientError(str(e))
        # else:
        # 更新令牌
        # await self.save_update_hash_data(self._account_key, field_name=session.account_id,
        #                                  hash_data=session.session_id, ex=LONG_EXPIRED)

    async def get_session(self, session_id: str, ex: int = SESSION_EXPIRED, cls_flag: bool = True,
                          load_responses: bool = False) -> Union[Session, Dict[str, str]]:
        """
        获取session
        Args:
            session_id: session id
            ex: 过期时间，单位秒
            cls_flag: 是否返回session的类实例
            load_responses: 结果的键值是否进行load
        Returns:

        """

        try:
            session_data = await self.redis_db.hgetall(session_id)
            if not session_data:
                raise RedisClientError("not found session, session_id={}".format(session_id))

            if not await self.redis_db.expire(session_id, ex):
                aelog.error("set session expire failed, session_id={}".format(session_id))
        except RedisError as e:
            aelog.exception("get session error: {}, {}".format(session_id, e))
            raise RedisClientError(e)
        else:
            # 返回的键值对是否做load
            if load_responses:
                hash_data = {}
                for hash_key, hash_val in session_data.items():
                    async with async_ignore_error():
                        hash_val = ujson.loads(hash_val)
                    hash_data[hash_key] = hash_val
                session_data = hash_data

            if cls_flag:
                return Session(session_data.pop('account_id'), session_id=session_data.pop('session_id'),
                               org_id=session_data.pop("org_id"), role_id=session_data.pop("role_id"),
                               permission_id=session_data.pop("permission_id"), **session_data)
            else:
                return session_data

    async def verify(self, session_id: str) -> Session:
        """
        校验session，主要用于登录校验
        Args:
            session_id
        Returns:

        """
        try:
            session = await self.get_session(session_id)
        except RedisClientError as e:
            raise RedisClientError(str(e))
        else:
            if not session:
                raise RedisClientError("invalid session_id, session_id={}".format(session_id))
            return session

    async def save_update_hash_data(self, name: str, hash_data: Dict, field_name: str = None, ex: int = EXPIRED,
                                    dump_responses: bool = False) -> str:
        """
        获取hash对象field_name对应的值
        Args:
            name: redis hash key的名称
            field_name: 保存的hash mapping 中的某个字段
            hash_data: 获取的hash对象中属性的名称
            ex: 过期时间，单位秒
            dump_responses: 是否对每个键值进行dump
        Returns:
            反序列化对象
        """
        if field_name is None and not isinstance(hash_data, MutableMapping):
            raise ValueError("hash data error, must be MutableMapping.")

        try:
            if not field_name:
                # 是否对每个键值进行dump
                if dump_responses:
                    rs_data = {}
                    for hash_key, hash_val in hash_data.items():
                        if not isinstance(hash_val, str):
                            async with async_ignore_error():
                                hash_val = ujson.dumps(hash_val)
                        rs_data[hash_key] = hash_val
                    hash_data = rs_data

                if not await self.redis_db.hmset(name, hash_data):
                    raise RedisClientError("save hash data mapping failed, session_id={}".format(name))
            else:
                hash_data = hash_data if isinstance(hash_data, str) else ujson.dumps(hash_data)
                await self.redis_db.hset(name, field_name, hash_data)

            if not await self.redis_db.expire(name, ex):
                aelog.error("set hash data expire failed, session_id={}".format(name))
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return name

    async def get_hash_data(self, name: str, field_name: str = None, ex: int = EXPIRED,
                            load_responses: bool = False) -> Dict:
        """
        获取hash对象field_name对应的值
        Args:
            name: redis hash key的名称
            field_name: 获取的hash对象中属性的名称
            ex: 过期时间，单位秒
            load_responses: 结果的键值是否进行load
        Returns:
            反序列化对象
        """
        try:
            if field_name:
                hash_data = await self.redis_db.hget(name, field_name)
                # 返回的键值对是否做load
                if load_responses:
                    async with async_ignore_error():
                        hash_data = ujson.loads(hash_data)
            else:
                hash_data = await self.redis_db.hgetall(name)
                # 返回的键值对是否做load
                if load_responses:
                    rs_data = {}
                    for hash_key, hash_val in hash_data.items():
                        async with async_ignore_error():
                            hash_val = ujson.loads(hash_val)
                        rs_data[hash_key] = hash_val
                    hash_data = rs_data
            if not hash_data:
                raise RedisClientError("not found hash data, name={}, field_name={}".format(name, field_name))

            if not await self.redis_db.expire(name, ex):
                aelog.error("set expire failed, name={}".format(name))
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return hash_data

    async def get_list_data(self, name: str, start: int = 0, end: int = -1, ex: int = EXPIRED) -> List:
        """
        获取redis的列表中的数据
        Args:
            name: redis key的名称
            start: 获取数据的起始位置,默认列表的第一个值
            end: 获取数据的结束位置，默认列表的最后一个值
            ex: 过期时间，单位秒
        Returns:

        """
        try:
            data = await self.redis_db.lrange(name, start=start, end=end)
            if not await self.redis_db.expire(name, ex):
                aelog.error("set expire failed, name={}".format(name))
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return data

    async def save_list_data(self, name: str, list_data: Union[List, str], save_to_left: bool = True,
                             ex: int = EXPIRED) -> str:
        """
        保存数据到redis的列表中
        Args:
            name: redis key的名称
            list_data: 保存的值,可以是单个值也可以是元祖
            save_to_left: 是否保存到列表的左边，默认保存到左边
            ex: 过期时间，单位秒
        Returns:

        """
        list_data = (list_data,) if isinstance(list_data, str) else list_data
        try:
            if save_to_left:
                if not await self.redis_db.lpush(name, *list_data):
                    raise RedisClientError("lpush value to head failed.")
            else:
                if not await self.redis_db.rpush(name, *list_data):
                    raise RedisClientError("lpush value to tail failed.")
            if not await self.redis_db.expire(name, ex):
                aelog.error("set expire failed, name={}".format(name))
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return name

    async def save_update_usual_data(self, name: str, value: Any, ex: int = EXPIRED) -> str:
        """
        保存列表、映射对象为普通的字符串
        Args:
            name: redis key的名称
            value: 保存的值，可以是可序列化的任何职
            ex: 过期时间，单位秒
        Returns:

        """
        value = ujson.dumps(value) if not isinstance(value, str) else value
        try:
            if not await self.redis_db.set(name, value, ex):
                raise RedisClientError("set serializable value failed!")
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return name

    async def incrbynumber(self, name: str, amount: int = 1, ex: int = EXPIRED) -> str:
        """

        Args:

        Returns:

        """
        try:
            if isinstance(amount, int):
                if not await self.redis_db.incr(name, amount):
                    raise RedisClientError("Increments int value failed!")
            else:
                if not await self.redis_db.incrbyfloat(name, amount):
                    raise RedisClientError("Increments float value failed!")
            if not await self.redis_db.expire(name, ex):
                aelog.error("set expire failed, name={}".format(name))
        except RedisError as e:
            raise RedisClientError(str(e))
        else:
            return name

    async def get_usual_data(self, name: str, load_responses: bool = True, update_expire: bool = True,
                             ex: int = EXPIRED) -> Union[Dict, str]:
        """
        获取name对应的值
        Args:
            name: redis key的名称
            load_responses: 是否转码默认转码
            update_expire: 是否更新过期时间
            ex: 过期时间，单位秒
        Returns:
            反序列化对象
        """
        data = await self.redis_db.get(name)

        if data is not None and update_expire:  # 保证key存在时设置过期时间
            if not await self.redis_db.expire(name, ex):
                aelog.error("set expire failed, name={}".format(name))

        if load_responses:
            async with async_ignore_error():
                data = ujson.loads(data)

        return data

    async def is_exist_key(self, name: str) -> bool:
        """
        判断redis key是否存在
        Args:
            name: redis key的名称
        Returns:

        """
        return await self.redis_db.exists(name)

    async def delete_keys(self, names: List[str]) -> NoReturn:
        """
        删除一个或多个redis key
        Args:
            names: redis key的名称
        Returns:

        """
        names = (names,) if isinstance(names, str) else names
        if not await self.redis_db.delete(*names):
            aelog.error("Delete redis keys failed {}.".format(*names))

    async def get_keys(self, pattern_name: str) -> List:
        """
        根据正则表达式获取redis的keys
        Args:
            pattern_name:正则表达式的名称
        Returns:

        """
        return await self.redis_db.keys(pattern_name)
